[Quant][fx][resubmit] Remove dequant-quant around getitem

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83172

Summary: https://github.com/pytorch/pytorch/issues/82480 saw
unnecessary dequant-quant pairs around the getitem op, which led
to significant slowdowns. This commit simply removes this pair in
the lowering step, since getitem already handles quantized inputs.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_getitem_no_dequant_quant

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/82480

Differential Revision: [D38657297](https://our.internmc.facebook.com/intern/diff/D38657297)

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel @ezyang @SherlockNoMad @soumith @EikanWang @wenzhe-nrv

andrewor14 committed:
[Quant][fx][resubmit] Remove dequant-quant around getitem

Summary: https://github.com/pytorch/pytorch/issues/82480 saw
unnecessary dequant-quant pairs around the getitem op, which led
to significant slowdowns. This commit simply removes this pair in
the lowering step, since getitem already handles quantized inputs.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_getitem_no_dequant_quant

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/82480

[ghstack-poisoned]

jerryzh168 commented:
was the internal test error resolved?

andrewor14 commented:
I'm still trying to reproduce it. Making a PR first so I can import

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38657297).

andrewor14 committed:
Update on "[Quant][fx][resubmit] Remove dequant-quant around getitem"


Summary: https://github.com/pytorch/pytorch/issues/82480 saw
unnecessary dequant-quant pairs around the getitem op, which led
to significant slowdowns. This commit simply removes this pair in
the lowering step, since getitem already handles quantized inputs.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_getitem_no_dequant_quant

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/82480

Differential Revision: [D38657297](https://our.internmc.facebook.com/intern/diff/D38657297)

[ghstack-poisoned]

jerryzh168 commented:
I think getitem support is explicitly skipped right now: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/prepare.py#L402-L408, maybe we can enable this instead?

andrewor14 commented:
> I think getitem support is explicitly skipped right now: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/prepare.py#L402-L408, maybe we can enable this instead?

I remember this broke some tests that were difficult to fix in the original PR. I think it had to do with `getitem` being used for dict inputs. Since this is a resubmit maybe we should revisit this in the future?

andrewor14 committed:
Update on "[Quant][fx][resubmit] Remove dequant-quant around getitem"


Summary: https://github.com/pytorch/pytorch/issues/82480 saw
unnecessary dequant-quant pairs around the getitem op, which led
to significant slowdowns. This commit simply removes this pair in
the lowering step, since getitem already handles quantized inputs.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_getitem_no_dequant_quant

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/82480

Differential Revision: [D38657297](https://our.internmc.facebook.com/intern/diff/D38657297)

[ghstack-poisoned]

andrewor14 commented:
> I think getitem support is explicitly skipped right now: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/prepare.py#L402-L408, maybe we can enable this instead?

As discussed offline, we cannot enable this without breaking the dict/list of tensors use case. I added a detailed explanation in the comments for context.

andrewor14 committed:
Update on "[Quant][fx][resubmit] Remove dequant-quant around getitem"


Summary: https://github.com/pytorch/pytorch/issues/82480 saw
unnecessary dequant-quant pairs around the getitem op, which led
to significant slowdowns. This commit simply removes this pair in
the lowering step, since getitem already handles quantized inputs.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_getitem_no_dequant_quant

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/82480

Differential Revision: [D38657297](https://our.internmc.facebook.com/intern/diff/D38657297)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38657297).

jerryzh168 reviewed (approved):
looks great, thanks!

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -1377,7 +1377,35 @@ def insert_observers_for_model(
 
+        # Second pass: Look for getitem nodes and make the input and output observers the same.
         #
+        # Context: There are two use cases of getitem:
```
should we move this to the doc for `_make_getitem_share_input_output_observers`?

andrewor14 committed:
Update on "[Quant][fx][resubmit] Remove dequant-quant around getitem"


Summary: https://github.com/pytorch/pytorch/issues/82480 saw
unnecessary dequant-quant pairs around the getitem op, which led
to significant slowdowns. This commit simply removes this pair in
the lowering step, since getitem already handles quantized inputs.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_getitem_no_dequant_quant

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/82480

Differential Revision: [D38657297](https://our.internmc.facebook.com/intern/diff/D38657297)

cc ezyang SherlockNoMad soumith EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38657297).

andrewor14 committed:
Update on "[Quant][fx][resubmit] Remove dequant-quant around getitem"


Summary: https://github.com/pytorch/pytorch/issues/82480 saw
unnecessary dequant-quant pairs around the getitem op, which led
to significant slowdowns. This commit simply removes this pair in
the lowering step, since getitem already handles quantized inputs.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_getitem_no_dequant_quant

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/82480

Differential Revision: [D38657297](https://our.internmc.facebook.com/intern/diff/D38657297)

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx][resubmit] Remove dequant-quant around getitem"


Summary: https://github.com/pytorch/pytorch/issues/82480 saw
unnecessary dequant-quant pairs around the getitem op, which led
to significant slowdowns. This commit simply removes this pair in
the lowering step, since getitem already handles quantized inputs.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_getitem_no_dequant_quant

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/82480

Differential Revision: [D38657297](https://our.internmc.facebook.com/intern/diff/D38657297)

[ghstack-poisoned]
